From a2a5d91a7d78d5194d5ffbac2393ab468ef47477 Mon Sep 17 00:00:00 2001
From: Yu Qiaowei <cerf.yu@rock-chips.com>
Date: Mon, 26 Jan 2026 20:03:59 +0800
Subject: [PATCH 27/87] video: rockchip: rga3: fix bi-linear scale-down
 coefficient check

Ensure src/dst last point mapping is confirmed after recalculating
final_steps.

Fixes: 971dce993d30 ("video: rockchip: rga3: fix bi-linear scaled down causing timeout")

Signed-off-by: Yu Qiaowei <cerf.yu@rock-chips.com>
Change-Id: I3e71a7867176e27a3de12c02e305a485e65c01fc
(cherry picked from commit 864c7565ca65dd48c2573ccb81f8d178d148d399)
---
 drivers/video/rockchip/rga3/rga2_reg_info.c | 41 ++++++++++++++-------
 1 file changed, 28 insertions(+), 13 deletions(-)

diff --git a/drivers/video/rockchip/rga3/rga2_reg_info.c b/drivers/video/rockchip/rga3/rga2_reg_info.c
index 236dfa4e7d054..24cdae11628ab 100644
--- a/drivers/video/rockchip/rga3/rga2_reg_info.c
+++ b/drivers/video/rockchip/rga3/rga2_reg_info.c
@@ -106,28 +106,43 @@ static void rga2_scale_down_bilinear_protect(u32 *param_fix, u32 *src_fix,
 {
 	int final_coor, final_diff, final_steps;
 
-	while (1) {
-		final_coor = offset + param * (dst - 1);
-		final_diff = (src - 1) * (1 << RGA2_BILINEAR_PREC) - final_coor;
+	final_steps = src - 1;
 
+	do {
 		/*
 		 *   The hardware requires that the last point of the dst map on
 		 * src must not exceed the range of src.
 		 */
-		if (final_diff <= 0)
+		do {
+			final_coor = offset + param * (dst - 1);
+			final_diff = final_steps * (1 << RGA2_BILINEAR_PREC) - final_coor;
+
+			if (final_diff <= 0 && param > 0)
+				param = param - 1;
+			else
+				break;
+		} while (1);
+
+		/*
+		 *   The hardware requires that the last point of dst mapping on
+		 * src be between the last two points of each row/column, so
+		 * actual width/height needs to be modified.
+		 */
+		final_steps = (final_coor & ((1 << RGA2_BILINEAR_PREC) - 1)) ?
+				((final_coor >> RGA2_BILINEAR_PREC) + 1) :
+				(final_coor >> RGA2_BILINEAR_PREC);
+
+		/*   After modifying final_steps, ensure once more that the last
+		 * point of the dst map on src must not exceed the range of src.
+		 */
+		final_diff = final_steps * (1 << RGA2_BILINEAR_PREC) - final_coor;
+		if (final_diff <= 0 && param > 0)
 			param = param - 1;
 		else
 			break;
-	}
+	} while (1);
 
-	/*
-	 *   The hardware requires that the last point of dst mapping on
-	 * src be between the last two points of each row/column, so
-	 * actual width/height needs to be modified.
-	 */
-	final_steps = (final_coor & ((1 << RGA2_BILINEAR_PREC) - 1)) ?
-		((final_coor >> RGA2_BILINEAR_PREC) + 1) :
-		(final_coor >> RGA2_BILINEAR_PREC);
+	WARN_ON(param == 0);
 
 	*param_fix = param;
 	*src_fix = final_steps + 1;
